### PR TITLE
Issue #249 Add Instance IP to host /etc/resolv.conf

### PR DESCRIPTION
**Pull request by praveenkumar**
This will allow us to add instance IP on fly so user doesn't
need to do it manually.

As preflight check we are also changing the permission for
host /etc/resolv.conf file to provide RW access to current user.

### CODE REVIEW

**Comment by praveenkumar** [comment]
retest this please.

**Comment by praveenkumar** [comment]
retest this please

File: pkg/crc/preflight/preflight_darwin.go [hunk 1]
```diff
@@ -23,6 +23,12 @@ func StartPreflightChecks() {
 		"Checking file permissions for resolver",
 		config.GetBool(cmdConfig.WarnCheckResolverFilePermissions.Name),
 	)
+
+	preflightCheckSucceedsOrFails(config.GetBool(cmdConfig.SkipCheckResolvConfFilePermissions.Name),
+		checkResolvConfFilePermissions,
+		"Checking file permissions for resolv.conf",
```
Comment (anjannath) [inline pkg/crc/preflight/preflight_darwin.go]: We should put the full path of the file `/etc/resolv.conf` and `/etc/resolver/testing`

File: pkg/crc/services/dns/dns_darwin.go [hunk 2]
```diff
@@ -28,8 +28,15 @@ type resolverFileValues struct {
...
-	success, err := createResolverFile(serviceConfig.IP, filepath.Join("/", "etc", "resolver", serviceConfig.BundleMetadata.ClusterInfo.BaseDomain))
+	// Update resolv.conf file for host
+	success, err := updateResolvConfFile(serviceConfig.IP, filepath.Join("/", "etc", "resolv.conf"))
+	if !success {
+		result.Success = success
+		return *result, err
+	}
+
+	// Write resolver config to host
+	success, err = createResolverFile(serviceConfig.IP, filepath.Join("/", "etc", "resolver", serviceConfig.BundleMetadata.ClusterInfo.BaseDomain))
```
Comment (anjannath) [inline pkg/crc/services/dns/dns_darwin.go]: We can update the `resolverFile` variable with this path

**Comment by gbraad** [comment]
Will review ASAP

File: pkg/crc/network/template.go [hunk 3]
```diff
@@ -7,13 +7,13 @@ import (
 
 const (
 	resolvFileTemplate = `# Generated by CRC
-search {{ range .SearchDomains }}{{ .Domain }}{{ end }}
+{{ range .SearchDomains }}search {{ .Domain }}{{ end }}
 {{ range .NameServers }}nameserver {{ .IPAddress }}
```
Comment (cfergeau) [inline pkg/crc/network/template.go]: Why is this changed?

File: pkg/crc/preflight/preflight_checks_darwin.go [hunk 4]
```diff
@@ -146,3 +147,27 @@ func fixOcBinaryCached() (bool, error) {
...
+		return false, err
+	}
+
+	stdOut, stdErr, err := crcos.RunWithPrivilege("chown", currentUser.Username, resolvFile)
+	if err != nil {
+		return false, fmt.Errorf("Unable to change ownership of the resolver file: %s %v: %s", stdOut, err, stdErr)
+	}
+
+	return true, nil
+}
```
Comment (cfergeau) [inline pkg/crc/preflight/preflight_checks_darwin.go]: You could introduce 2 helpers to share the code with `checkResolverFilePermissions` and `fixResolverFilePermissions`

File: pkg/crc/services/dns/dns_darwin.go [hunk 5]
```diff
@@ -28,8 +28,15 @@ type resolverFileValues struct {
 }
 
 func runPostStartForOS(serviceConfig services.ServicePostStartConfig, result *services.ServicePostStartResult) (services.ServicePostStartResult, error) {
-	// Write resolver file
-	success, err := createResolverFile(serviceConfig.IP, filepath.Join("/", "etc", "resolver", serviceConfig.BundleMetadata.ClusterInfo.BaseDomain))
+	// Update resolv.conf file for host
+	success, err := updateResolvConfFile(serviceConfig.IP, filepath.Join("/", "etc", "resolv.conf"))
```
Comment (cfergeau) [inline pkg/crc/services/dns/dns_darwin.go]: Is there a bug filed against `oc` to ask them to use the native macos APIs? (though I suspect this comes from go stdlib).

File: pkg/crc/network/template.go (hunk 3 above)
Comment (praveenkumar) [reply pkg/crc/network/template.go]: @cfergeau this is changed because if there is no search in the `/etc/resolv.conf` then end result after processing the template will be `search ` without anything and that will cause issue for command like dig to parse it.

File: pkg/crc/preflight/preflight_checks_darwin.go (hunk 4 above)
Comment (praveenkumar) [reply pkg/crc/preflight/preflight_checks_darwin.go]: Yes, this refactor I can do 👍

File: pkg/crc/services/dns/dns_darwin.go (hunk 5 above)
Comment (praveenkumar) [reply pkg/crc/services/dns/dns_darwin.go]: No, I didn't file it but as you said it is going to come from go stdlib, not sure if filling the issue there will make any sense.

Comment (cfergeau) [reply pkg/crc/services/dns/dns_darwin.go]: On the other hand, adding a workaround in our code without noticing other parties of the bug means this will never get fixed in the right place :(

File: pkg/crc/network/template.go (hunk 3 above)
Comment (cfergeau) [reply pkg/crc/network/template.go]: Ah I see. This could almost go in a separate commit as this is an unrelated bugfix.. Is this also behaving as expected when multiple search domains are in use?

Comment (praveenkumar) [reply pkg/crc/network/template.go]: yes now it behave as expected even multiple search domains are in use.

File: pkg/crc/services/dns/dns_darwin.go (hunk 5 above)
Comment (praveenkumar) [reply pkg/crc/services/dns/dns_darwin.go]: @cfergeau it is not only the `oc` even dig doesn't honour the resolver for mac, but I do get your point will try to create an issue on origin side.

Comment (cfergeau) [reply pkg/crc/services/dns/dns_darwin.go]: On my machine, it's only `oc` and `dig` which are using `resolv.conf`, `ping` for example does not seem to rely on it.

File: pkg/crc/preflight/preflight_checks_darwin.go [hunk 6]
```diff
@@ -146,3 +125,40 @@ func fixOcBinaryCached() (bool, error) {
...
+func addFileWritePermissionToUser(filename string) (bool, error) {
+	logging.DebugF("Making %s readable/writable by the current user", filename)
+	currentUser, err := user.Current()
+	if err != nil {
+		return false, err
+	}
+
+	stdOut, stdErr, err := crcos.RunWithPrivilege("chown", currentUser.Username, filename)
+	if err != nil {
+		return false, fmt.Errorf("Unable to change ownership of the resolver file: %s %v: %s", stdOut, err, stdErr)
```
Comment (cfergeau) [inline pkg/crc/preflight/preflight_checks_darwin.go]: `resolver` should be removed or replaced with `filename`

File: pkg/crc/preflight/preflight_checks_darwin.go [hunk 7]
```diff
@@ -146,3 +125,40 @@ func fixOcBinaryCached() (bool, error) {
...
+	}
+
+	err = os.Chmod(filename, 0644)
+	if err != nil {
+		return false, fmt.Errorf("Unable to change permissions of the resolver file: %s %v: %s", stdOut, err, stdErr)
```
Comment (cfergeau) [inline pkg/crc/preflight/preflight_checks_darwin.go]: Same comment here
